pkg/sshagent/sshagent.go

```diff
@@ -0,0 +1,225 @@
+	}
+	return newAgentServerSocket(source.Socket)
+}
+
+// newAgentServerKeyring creates an new agent from scratch and adds keys
+func newAgentServerKeyring(keys []interface{}) (*AgentServer, error) {
+	a := agent.NewKeyring()
+	for _, k := range keys {
+		if err := a.Add(agent.AddedKey{PrivateKey: k}); err != nil {
+			return nil, errors.Wrap(err, "failed to create ssh agent")
+		}
+	}
+	return &AgentServer{
+		agent:    a,
+		shutdown: make(chan bool, 1),
+	}, nil
+}
+
+// newAgentServerSocket creates an new agent from an existing agent on the host
+func newAgentServerSocket(socketPath string) (*AgentServer, error) {
+	conn, err := net.Dial("unix", socketPath)
+	if err != nil {
+		return nil, err
+	}
+	a := &readOnlyAgent{agent.NewClient(conn)}
+
+	return &AgentServer{
+		agent:    a,
+		conn:     &conn,
+		shutdown: make(chan bool, 1),
+	}, nil
+
+}
+
+// Serve starts the SSH agent on the host and returns the path of the socket where the agent is serving
+func (a *AgentServer) Serve(processLabel string) (string, error) {
+	err := selinux.SetSocketLabel(processLabel)
+	if err != nil {
+		return "", err
+	}
+	serveDir, err := ioutil.TempDir("", ".buildah-ssh-sock")
+	if err != nil {
+		return "", err
+	}
+	servePath := filepath.Join(serveDir, "ssh_auth_sock")
+	a.serveDir = serveDir
+	a.servePath = servePath
+	listener, err := net.Listen("unix", servePath)
+	if err != nil {
+		return "", err
+	}
+	err = selinux.SetSocketLabel("")
+	if err != nil {
+		return "", err
+	}
+	a.listener = listener
+
+	go func() {
+		for {
+			//listener.Accept blocks
+			c, err := listener.Accept()
+			if err != nil {
+				select {
+				case <-a.shutdown:
+					return
+				default:
+					logrus.Errorf("error accepting SSH connection: %v", err)
+					continue
+				}
+			}
+			a.wg.Add(1)
+			go func() {
+				// agent.ServeAgent will only ever return with error,
+				err := agent.ServeAgent(a.agent, c)
+				if err != io.EOF {
+					logrus.Errorf("error serving agent: %v", err)
+				}
+				a.wg.Done()
+			}()
+			time.Sleep(5 * time.Millisecond)
+			c.Close()
+		}
+	}()
+	return a.servePath, nil
+}
+
+// Shutdown shuts down the agent and closes the socket
+func (a *AgentServer) Shutdown() error {
+	if a.listener != nil {
+		a.shutdown <- true
+		a.listener.Close()
+	}
+	if a.conn != nil {
+		conn := *a.conn
+		conn.Close()
+	}
+	a.wg.Wait()
+	err := os.RemoveAll(a.serveDir)
+	if err != nil {
+		return err
+	}
+	a.serveDir = ""
+	a.servePath = ""
+	return nil
+}
+
+// ServePath returns the path where the agent is serving
+func (a *AgentServer) ServePath() string {
+	return a.servePath
+}
+
+// readOnlyAgent implemetnts the agent.Agent interface
+// readOnlyAgent allows reads only to prevent keys from being added from the build to the forwarded ssh agent on the host
+type readOnlyAgent struct {
+	agent.Agent
+}
+
+func (a *readOnlyAgent) Add(_ agent.AddedKey) error {
+	return errors.Errorf("adding new keys not allowed by buildkit")
```

Comment:
buildkit?  Shouldn't this say buildah?
If you copied this code out of buildkit, you should give some credit to them in the header.

Reply:
Theoretically should these be errors.New()
Ditto for all of the other references to buildkit and errors.Errorf.

Reply:
Does something like 
// readOnlyAgent and its functions originally from moby/buildkit 
work?